ospfd/ospf_asbr.c

```diff
@@ -305,3 +307,764 @@ void ospf_redistribute_withdraw(struct ospf *ospf, uint8_t type,
+{
+	struct route_node *rn = NULL;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("Unset the origination bit for all aggregator");
+
+	for (rn = route_top(ospf->rt_aggr_tbl); rn; rn = route_next(rn)) {
+		if (!rn->info)
+			continue;
+
+		struct ospf_external_aggr_rt *aggr = rn->info;
+
+		UNSET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+	}
+}
+
+static void ospf_delete_all_marked_aggregators(struct ospf *ospf)
+{
+	struct route_node *rn = NULL;
+
+	/* Loop through all the aggregators, Delete all aggregators
+	 * which are marked as DELETE. Set action to NONE for remaining
+	 * aggregators
+	 */
+	for (rn = route_top(ospf->rt_aggr_tbl); rn; rn = route_next(rn))
+		if (rn->info) {
+			struct ospf_external_aggr_rt *aggr = rn->info;
+
+			if (aggr->action != OSPF_ROUTE_AGGR_DEL) {
+				aggr->action = OSPF_ROUTE_AGGR_NONE;
+				continue;
+			}
+			ospf_external_aggr_delete(ospf, rn);
+			ospf_external_aggrigator_free(aggr);
+		}
+}
+
+static void ospf_handle_aggregated_exnl_rt(struct ospf *ospf,
+					   struct ospf_external_aggr_rt *aggr,
+					   struct external_info *ei)
+{
+	struct ospf_lsa *lsa;
+	struct as_external_lsa *al;
+	struct in_addr mask;
+
+	/* Handling the case where the external route prefix
+	 * and aggegate prefix is same
+	 * If same dont flush the originated external LSA.
+	 */
+	if (prefix_same((struct prefix *)&aggr->p, (struct prefix *)&ei->p)) {
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug(
+				"%s: External Route prefix same as Aggregator(%s/%d), so dont flush.",
+				__func__, inet_ntoa(ei->p.prefix),
+				ei->p.prefixlen);
+		return;
+	}
+
+	lsa = ospf_external_info_find_lsa(ospf, &ei->p);
+	if (lsa) {
+		al = (struct as_external_lsa *)lsa->data;
+		masklen2ip(ei->p.prefixlen, &mask);
+
+		if (mask.s_addr != al->mask.s_addr)
+			return;
+
+		ospf_external_lsa_flush(ospf, ei->type, &ei->p, 0);
+	}
+}
+
+static void ospf_handle_exnl_rt_after_aggr_del(struct ospf *ospf,
+					       struct external_info *ei)
+{
+	struct ospf_lsa *lsa;
+
+	/* Process only marked external routes.
+	 * These routes were part of a deleted
+	 * aggregator.So, originate now.
+	 */
+	if (!ei->to_be_processed)
+		return;
+
+	ei->to_be_processed = false;
+
+	lsa = ospf_external_info_find_lsa(ospf, &ei->p);
+
+	if (lsa)
+		ospf_external_lsa_refresh(ospf, lsa, ei, LSA_REFRESH_FORCE, 0);
+	else {
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug("%s: Originate external route(%s/%d)",
+				   __func__, inet_ntoa(ei->p.prefix),
+				   ei->p.prefixlen);
+
+		ospf_external_lsa_originate(ospf, ei);
+	}
+}
+
+static void ospf_handle_external_aggr_add(struct ospf *ospf)
+{
+	struct external_info *ei;
+	struct route_node *rn = NULL;
+	struct route_table *rt = NULL;
+	int type = 0;
+
+	/* Delete all the aggregators which are marked as
+	 * OSPF_ROUTE_AGGR_DEL.
+	 */
+	ospf_delete_all_marked_aggregators(ospf);
+
+	for (type = 0; type <= ZEBRA_ROUTE_MAX; type++) {
+		struct list *ext_list;
```

Comment:
If I understood the question correctly , yes we have to loop for all route types . 
Since we have to check for all redistributed route types(static, connected , bgp ... etc) in ospf and apply the aggregation.